### ISSUE TITLE
Biblatex 3.15a & dev, babel, and extra space before \smartof

### ISSUE THREAD

**Issue by dfussner** [plk/biblatex#1069]
When I use `babel` with French  or Catalan as the main language, there is currently
always an extra space before the "de" or "d'" produced by
`\smartof` before a title. `Polyglossia` produces correct output.

Here's a MWE:

```
\documentclass[12pt]{article}
\usepackage{fontspec}
\usepackage{csquotes}
% \usepackage{polyglossia}
% \setdefaultlanguage{french}
% \setotherlanguage[variant=american]{english}
\usepackage[american,french]{babel}
\usepackage[style=verbose-inote,backend=biber]{biblatex}
%\renewrobustcmd*{\smartof}{\unspace\global\toggletrue{smartof}}
\begin{filecontents}[overwrite]{\jobname.bib}
@book{knuth:ct:a,
  author       = {Knuth, Donald E.},
  title        = {The Passing of an Illusion},
  date         = {1984},
  maintitle    = {Computers \& Typesetting},
  related      = {knuth:ct:t},
  relatedtype  = {reprintfrom},
  volume       = {A},
  publisher    = {Addison-Wesley},
  location     = {Reading, Mass.}
}
@book{knuth:ct:t,
  author       = {Knuth, Donald E.},
  maintitle    = {The Passing of an Illusion},
  date         = {1984},
  translator   = {Any Name},
  publisher    = {Addison-Wesley},
  location     = {Reading, Mass.}
}
\end{filecontents}
\addbibresource{\jobname.bib}
\begin{document}
\thispagestyle{empty}

\noindent\cite{knuth:ct:a}.

\printbibliography
\end{document}

```
Here's what it looks like on my system in French:
![mwe17](https://user-images.githubusercontent.com/5174434/98691432-ee343500-2365-11eb-9de6-1d987ea33b80.png)

And here's the Catalan output:
![mwe18](https://user-images.githubusercontent.com/5174434/98692897-8848ad00-2367-11eb-91a2-30369e3a3440.png)

As you can see, the spacing before a name appears OK in the Catalan example, but is 
wrong before a title. Perhaps something to do with `\mkrelatedstring` (now `\mkrelatedstringtext`)
 in `title:hook`?

You can remove the space before `\smartof` in the actual `bibstring` and
it looks fine before name and title with both `babel` and `polyglossia`. Similarly,
you can redefine `\smartof` (as in the commented line above) to include an
`\unspace` and it again looks fine with both packages. I haven't been
able to determine when this ceased to work right. (At a guess, it's
something in `babel`, as the `\smartof` code has been fairly stable for
years, but I hoped you could find a better workaround than mine.)

Many thanks.

**Comment by moewew** [comment]
I really don't understand why `polyglossia` gives the desired result, but I think the correct solution is to remove the space before `\smartof` in the bibstring here, since the space is already implied by other field formats.

**Comment by dfussner** [comment]
I pulled the dev version this morning, and unfortunately I think the issue is a bit more general than just that one 'reprintfrom' `bibstring`. If, in the MWE above, you now replace `relatedtype` 'reprintfrom' with 'bytranslator' in the first entry, you get the same extra space using `babel`, like so (this is Catalan):
![mwe19](https://user-images.githubusercontent.com/5174434/98793515-ede87800-23ff-11eb-8f9b-7cec8d090cfc.png)
 
The change from de to d' also doesn't occur here. That same string worked fine when it wasn't the `relatedstring`, as above, so I guess there's an extra space always implied by the `related` mechanism? At least when using `babel` -- `polyglossia` doesn't show the extra space, but it, too, produces de instead of d' in this context.

**Comment by dfussner** [comment]
For what it's worth, here's a patch against dev tip that seems here to sort out all the issues I've seen, both with `babel` and with `polyglossia`. It reverts 602407e, as I was worried that those spaces might be useful somewhere (though I can't work out where, exactly).

```
diff --git a/tex/latex/biblatex/biblatex.def b/tex/latex/biblatex/biblatex.def
index ab668444..06921122 100644
--- a/tex/latex/biblatex/biblatex.def
+++ b/tex/latex/biblatex/biblatex.def
@@ -589,8 +589,10 @@
 \DeclareFieldFormat{related:multivolume}{#1}
 \DeclareFieldFormat{related:origpubin}{\mkbibparens{#1}}
 \DeclareFieldFormat{related:origpubas}{\mkbibparens{#1}}
-\DeclareFieldFormat{relatedstring:default}{#1\printunit{\relatedpunct}}
-\DeclareFieldFormat{relatedstring:reprintfrom}{#1\addspace}
+\DeclareFieldFormat{relatedstring:default}{%
+  #1\iftoggle{smartof}{}{\printunit{\relatedpunct}}}
+\DeclareFieldFormat{relatedstring:reprintfrom}{%
+  #1\iftoggle{smartof}{}{\addspace}}
 
 % Generic formats for \printtext and \printfield
 
@@ -2988,12 +2990,14 @@
        \renewbibmacro*{pageref}{}}
       {\thefield{entrytype}}}}
 
+\newbibmacro*{bytrans:hook}[1]{%
+  \ifnumequal{\value{listcount}}{1}%
+  {\mkrelatedstringtext{\lbx@initnamehook{#1}}}%
+  {}}%
+
 \newbibmacro*{related:bytranslator}[1]{%
   \entrydata{#1}{%
-    \renewbibmacro*{name:hook}[1]{%
-      \ifnumequal{\value{listcount}}{1}
-        {\mkrelatedstringtext{\lbx@initnamehook{#1}}}
-        {}}%
+    \letbibmacro*{name:hook}{bytrans:hook}%
     \printnames[bytranslator]{translator}%
     \setunit*{\addspace\bibstring[\mkrelatedstringtext]{astitle}\space}%
     \usebibmacro{title}%
diff --git a/tex/latex/biblatex/lbx/catalan.lbx b/tex/latex/biblatex/lbx/catalan.lbx
index d484f3d7..4f1878ac 100644
--- a/tex/latex/biblatex/lbx/catalan.lbx
+++ b/tex/latex/biblatex/lbx/catalan.lbx
@@ -375,7 +375,7 @@
   reprint          = {{reimpressi\'o}{reimpr\adddot}},
   reprintof        = {{reimpressi\'o \smartof}{reimpr\adddotspace\smartof}},
   reprintas        = {{reimpr\`es com}{reimpr\adddotspace com}},
-  reprintfrom      = {{reimpr\`es\smartof}{reimpr\adddot\smartof}},
+  reprintfrom      = {{reimpr\`es \smartof}{reimpr\adddotspace\smartof}},
   reviewof         = {{ressenya \smartof}{ress\adddotspace\smartof}},
   translationof    = {{traducci\'o \smartof}{trad\adddotspace\smartof}},
   translationas    = {{tradu\"{\i}t com}{trad\adddotspace com}},
diff --git a/tex/latex/biblatex/lbx/french.lbx b/tex/latex/biblatex/lbx/french.lbx
index 6354d793..d2e441b6 100644
--- a/tex/latex/biblatex/lbx/french.lbx
+++ b/tex/latex/biblatex/lbx/french.lbx
@@ -378,7 +378,7 @@
   reprint          = {{r\'eimpression}{r\'eimpr\adddot}},
   reprintof        = {{r\'eimpression de}{r\'eimpr\adddotspace de}},
   reprintas        = {{r\'eimpression sous le titre}{r\'eimpr\adddotspace sous le titre}},
-  reprintfrom      = {{r\'eimpression \`a partir\smartof}{r\'eimpr\adddotspace \`a part\adddot\smartof}},
+  reprintfrom      = {{r\'eimpression \`a partir \smartof}{r\'eimpr\adddotspace \`a part\adddotspace\smartof}},
   translationof    = {{traduction de}{trad\adddotspace de}},
   translationas    = {{traduit sous le titre}{trad\adddotspace sous le titre}},
   translationfrom  = {{traduit de}{trad\adddotspace de}},
```

**Comment by moewew** [comment]
The incorrect d'/de for `bytranslator` was caused by a typo. It could be resolved as in your fix, but the fix in https://github.com/plk/biblatex/commit/ce3cecb931901386cce292ba06f395349e5e4072 is a wee bit quicker.

---

I'll have to think about the spaces again.

**Comment by dfussner** [comment]
ce3cecb is much the better fix, thank you!

**Comment by moewew** [comment]
https://github.com/plk/biblatex/commit/28a0bd9ec4e660acf62fb6a4547361f7bb68f133 should fix the double spaces. Since the space is already implied by how `\smartof` is implemented we don't need it in the bibstrings.

**Comment by moewew** [comment]
Should hopefully be resolved in `biblatex` v3.16.
